aten/src/ATen/native/cpu/SumKernel.cpp

```diff
@@ -0,0 +1,304 @@
+  }
+}
+
+template <typename scalar_t>
+void scalar_inner_sum(
+    char * C10_RESTRICT data[2], int64_t in_strides[2], int64_t out_stride,
+    int64_t size0, int64_t size1) {
+  for (int64_t j = 0; j < size1; ++j) {
+    const auto *row_in = data[1] + j * in_strides[1];
+    scalar_t ans = row_sum<scalar_t>(row_in, in_strides[0], size0);
+    store_result(data[0], out_stride, j, ans);
+  }
+}
+
+template <typename scalar_t>
+void vectorized_outer_sum(
+    char * C10_RESTRICT data[2], int64_t inner_stride, int64_t out_stride,
+    int64_t size0, int64_t size1) {
+  using vec_t = Vec256<scalar_t>;
+  constexpr int64_t nrows = 4;
+  constexpr int64_t vec_stride = vec_t::size() * sizeof(scalar_t);
+
+  // Input is contiguous over the second (non-reduced) dimension
+  int64_t j = 0;
+  for (; j + nrows * vec_t::size() <= size1; j += nrows * vec_t::size()) {
+    const auto *row_in = data[1] + j * sizeof(scalar_t);
+    auto sums = multi_row_sum<vec_t, nrows>(row_in, inner_stride, vec_stride, size0);
+
+    for (int64_t i = 0; i < nrows; ++i) {
+      const int64_t base_idx = j + i * vec_t::size();
+
+      std::array<scalar_t, vec_t::size()> ans;
+      sums[i].store(ans.data());
+      store_result(data[0], out_stride, base_idx, ans);
+    }
+  }
+
+  for (; j + vec_t::size() <= size1; j += vec_t::size()) {
+    const auto *row_in = data[1] + j * sizeof(scalar_t);
+    const vec_t sums = row_sum<vec_t>(row_in, inner_stride, size0);
+
+    std::array<scalar_t, vec_t::size()> ans;
+    sums.store(ans.data());
+    store_result(data[0], out_stride, j, ans);
+  }
+
+  for (; j < size1; ++j) {
+    const auto *row_in = data[1] + j * sizeof(scalar_t);
+    scalar_t ans = row_sum<scalar_t>(row_in, inner_stride, size0);
+    store_result(data[0], out_stride, j, ans);
+  }
+}
+
+template <typename scalar_t>
+void scalar_outer_sum(
+    char * C10_RESTRICT data[2], int64_t in_strides[2], int64_t out_stride,
+    int64_t size0, int64_t size1) {
+
+  constexpr int64_t nrows = 4;
+  int64_t j = 0;
+  for (; j + (nrows - 1) < size1; j += nrows) {
+    const auto *row_in = data[1] + j * in_strides[1];
+    auto sums = multi_row_sum<scalar_t, nrows>(
+        row_in, in_strides[0], in_strides[1], size0);
+    store_result(data[0], out_stride, j, sums);
+  }
+
+  for (; j < size1; ++j) {
+    const auto *row_in = data[1] + j * in_strides[1];
+    scalar_t ans = row_sum<scalar_t>(row_in, in_strides[0], size0);
+    store_result(data[0], out_stride, j, ans);
+  }
+}
+
+void sum_kernel_impl(TensorIterator &iter) {
+  if (isIntegralType(iter.dtype(), /*includeBool=*/ true)) {
+    AT_DISPATCH_INTEGRAL_TYPES_AND(ScalarType::Bool, iter.dtype(), "sum_cpu",
+      [&] {
+        binary_kernel_reduce_vec(
+            iter, [=](scalar_t a, scalar_t b) -> scalar_t { return a + b; },
+            [=](Vec256<scalar_t> a, Vec256<scalar_t> b) { return a + b; });
+      });
+    return;
+  }
+
+  // Custom floating point sum for better accuracy
```

Comment:
There should be a high level description of what the algorithm does, similar to what you've already in the PR description, in the source code here, to make it easier for people to figure out what is going on.